Slack tool

EugeneLightsOn opened the pull request:
Slack Tool implementation

https://github.com/user-attachments/assets/d3829cac-198f-4120-941a-14df0185e1f4

**AI Description**

<!-- begin-generated-description -->

This PR adds a new Slack tool to the deployable all-in-one RAG application. The Slack tool allows users to search Slack for relevant document snippets.

The PR includes the following changes:

- A new file, `docs/custom_tool_guides/slack.md`, is added, which provides instructions for setting up the Slack tool.
- The `poetry.lock` file is updated to include the `slack-sdk` package, which is required for the Slack tool.
- The `pyproject.toml` file is updated to include the `slack-sdk` package as a dependency.
- The `src/backend/config/configuration.template.yaml` file is updated to include a new `slack` section, which specifies the user scopes required for the Slack tool.
- The `src/backend/config/secrets.template.yaml` file is updated to include a new `slack` section, which specifies the client ID and client secret required for the Slack tool.
- The `src/backend/config/settings.py` file is updated to include a new `SlackSettings` class, which is used to configure the Slack tool.
- The `src/backend/config/tools.py` file is updated to include new `SlackAuth` and `SlackTool` classes, which are used to authenticate and interact with the Slack tool.
- The `src/backend/tools/__init__.py` file is updated to include new `SlackAuth` and `SlackTool` classes, which are used to authenticate and interact with the Slack tool.
- The `src/backend/tools/base.py` file is updated to include a new `tool_auth_crud.delete_tool_auth` method, which is used to delete tool authentication information.
- The `src/backend/tools/slack/__init__.py` file is added, which imports the `SlackAuth`, `SlackTool`, and `SLACK_TOOL_ID` constants from the `backend.tools.slack` module.
- The `src/backend/tools/slack/auth.py` file is added, which defines the `SlackAuth` class, which is used to authenticate the Slack tool.
- The `src/backend/tools/slack/client.py` file is added, which defines the `SlackClient` class, which is used to interact with the Slack API.
- The `src/backend/tools/slack/constants.py` file is added, which defines the `SEARCH_LIMIT` and `SLACK_TOOL_ID` constants, which are used by the Slack tool.
- The `src/backend/tools/slack/tool.py` file is added, which defines the `SlackTool` class, which is used to search Slack for relevant document snippets.
- The `src/backend/tools/slack/utils.py` file is added, which defines the `SlackService` class, which is used to interact with the Slack API.
- The `src/interfaces/assistants_web/src/app/(main)/settings/Settings.tsx` file is updated to include a new `SlackConnection` component, which is used to display the Slack connection status.
- The `src/interfaces/assistants_web/src/components/AgentSettingsForm/StatusConnection.tsx` file is added, which defines the `StatusConnection` component, which is used to display the Slack connection status.
- The `src/interfaces/assistants_web/src/components/AgentSettingsForm/ToolsStep.tsx` file is updated to include a new `handleAuthButtonClick` prop, which is used to handle authentication for the Slack tool.
- The `src/interfaces/assistants_web/src/components/AgentSettingsForm/index.tsx` file is updated to include a new `handleAuthButtonClick` method, which is used to handle authentication for the Slack tool.
- The `src/interfaces/assistants_web/src/components/UI/Icon.tsx` file is updated to include a new `Slack` component, which is used to display the Slack icon.
- The `src/interfaces/assistants_web/src/constants/tools.ts` file is updated to include a new `TOOL_SLACK_ID` constant, which is used to identify the Slack tool.

<!-- end-generated-description -->

tianjing-li commented on docs/custom_tool_guides/slack.md:
```diff
@@ -0,0 +1,128 @@
+## 3. Set Up Environment Variables
+
+Then set the following environment variables. You can either set the below values in your `.env` file:
+
```
Small request: Can you switch the ordering of 3. so that we recommend updating `secrets.yaml` first, then offer `.env` as a fallback solution

EugeneLightsOn replied:
Thanks, I switched the order

tianjing-li commented on src/backend/tools/slack/auth.py:
```diff
@@ -0,0 +1,159 @@
+
+        if (
+            self.SLACK_CLIENT_ID is None
+            or self.SLACK_CLIENT_SECRET is None
```
Can use 
```
if any([
     A is None, 
     B is None,
]):
```

EugeneLightsOn replied:
Good point, thanks.

tianjing-li commented on src/backend/tools/slack/auth.py:
```diff
@@ -0,0 +1,159 @@
+
+        settings = Settings()
+        slack_settings = settings.tools.slack if settings.tools and settings.tools.slack else None
+        self.SLACK_CLIENT_ID = getattr(slack_settings, 'client_id', None)
```
Prefer: To maintain consistency across the codebase, can you fetch the config with

```
Settings().tools.slack.myconfig
```
For each variable instead? This will also set them as None automatically if it's not populated

EugeneLightsOn replied:
Pyright checks. We've discussed that

tianjing-li commented on src/backend/tools/slack/tool.py:
```diff
@@ -0,0 +1,55 @@
+
+
+class SlackTool(BaseTool):
+    """
```
Can you add all the object types it searches (eg Files/Messages)

EugeneLightsOn replied:
Ty, done

tianjing-li commented:
Great PR. Added a couple comments.

In addition:
Client-side, can we remove Slack as a tool for the base Agent
<img width="398" alt="image" src="https://github.com/user-attachments/assets/8c35c656-e952-4a49-81a6-2344bc45683f">

I also have some general questions for the tool -> service -> client architecture, I'll reach out in Slack

EugeneLightsOn commented:
> Great PR. Added a couple comments.
> 
> In addition: Client-side, can we remove Slack as a tool for the base Agent <img alt="image" width="398" src="https://private-user-images.githubusercontent.com/18335148/381274288-8c35c656-e952-4a49-81a6-2344bc45683f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyMzE1MDAsIm5iZiI6MTczMDIzMTIwMCwicGF0aCI6Ii8xODMzNTE0OC8zODEyNzQyODgtOGMzNWM2NTYtZTk1Mi00YTQ5LTgxYTYtMjM0NGJjNDU2ODNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI5VDE5NDY0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2ZmYxOTg3YTdhOWY3YzY1ZTA1YTg0ZGZjMWFiNDNjYTQ4ZWJhN2VhNTFkNWVkZjI4NTI5N2FkZGJkMGE1N2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YraqPxToSaLBkvURHBzSaHlV3ArftRPxLQP78PJ3SUg">
> 
> I also have some general questions for the tool -> service -> client architecture, I'll reach out in Slack

I've added a constant to manage which tools should be excluded from the base agent.

EugeneLightsOn commented on src/backend/tools/slack/tool.py:
```diff
@@ -0,0 +1,55 @@
+    def is_available(cls) -> bool:
+        settings = Settings()
+        slack_settings = settings.tools.slack if settings.tools and settings.tools.slack else None
+        cls.CLIENT_ID = getattr(slack_settings, 'client_id', None)
```
Yes, it's related to the Pyright checks. I'll refactor it after the Config Settings improvements are completed.
